### DISCUSSION TITLE
The path to API coverage

### DISCUSSION THREAD

**Discussion by pkese** [DiffSharp/DiffSharp#380] [General]
#### DiffSharp is currently providing rather limited APIs compared to Torch

> Tinkering with slightly more advanced models quickly leads to discovery   
> that many familiar Torch APIs are not yet available in DiffSharp.

Since I find DiffSharp source code reasonably understandable, I was thinking I could even add the missing APIs myself (linking Torch-backend to TorchSharp DLL), but ... the way DiffSharp is structured, I assume I'd also have to manually create a "from first principles" Reference implementation in order for things to compile and tests to pass.

Problem is that this is way more work, while not adding any practical benefit (I'm not going to use Reference anyway).

And that's discouraging me from investing my time.

So my thinking is, that maybe DiffSharp would get more velocity and community support as a project, if it pragmatically just focused on the Torch backend and let other backends lag behind in implementation coverage.  

What I'm thinking of is structuring the code in such a way that things would build and tests would not collapse in case some backends were incomplete.

Oh, and what could also help is adding a how-to document explaining how to pull a missing API from TorchSharp and add related tests.

Any opinions?

**Comment by dsyme** [comment]
@pkese Some thoughts

* You're right that requiring reference implementations of everything is not practical.  We have to include ways for people to simply work with TorchTensor where necessary and map that up to DiffSharp.Tensor.

* DiffSharp and TorchSharp should be considered part of the same ecosystem.  It's possible we could even make them nearly 100% compatible APIs in the end. We'd like people to think of TorchSharp as the core and DiffSharp as a particualr kind of value add.  

* *If you're only interested in plain tensors with first-order backward mode gradients then you should probably use TorchSharp* (from F# of course :-) ).  Any specific advantages that DiffSharp has (e.g. say better printing or nicer API) for that mode of work should ideally be made available in TorchSharp.  We don't want DiffSharp to be "nicer" than TorchSharp, though there may be differences.

* So, let's assume you need forward mode or 2nd order or nested differentiability. Often, there will be a core of tensor code which requires forward/2nd-order/nested differentiability, and the rest which doesn't (e.g. data loading, image manipulation and so on).   So you won't need full nested fwd/bwd differentiability for everything.  However for that core, we require nested-differentiable versions of all primitives.  This means `DiffSharp.Tensor` code like that in Tensor.fs and Model.fs and similar.

* Practically speaking you can raise `TorchTensor` functions to `DiffSharp.Tensor` functions through [the extension API](https://diffsharp.github.io/extensions.html). For each function you lift you can also include backward gradient derivatives implemented via TorchTensor `gradient`, and also spec out the fwd/bwd/nested if necessary.  You don't need to give a reference backend implementation.   We should write up how to do this (it involves casting from RawTensor to TorchRawTensor, and failing for ReferenceRawTensor).  

* Given how quickly TorchSharp has advanced, perhaps we should indeed add a `DiffSharp.Torch` which basically maps the entire TorchSharp API up into DiffSharp, filling in as much of the derivative logic as is easily possible (implemented via TorchSharp `gradient`).

**Reply by pkese** [reply]
Thanks.
I'll stick with TorchSharp since I'm only using simple gradients anyway.
